static/js/models/reopening-web-socket.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/**
+ * Mirrors the WS API but automatically reopens on connection
+ * interruption/error
+ */
+class ReopeningWebSocket {
+  constructor(href) {
+    this.href = href;
+    this.ws = null;
+    this.closing = false;
+    this.opening = false;
+
+    this.onOpen = this.onOpen.bind(this);
+    this.onMessage = this.onMessage.bind(this);
+    this.reopen = this.reopen.bind(this);
+
+    this.listeners = {
+      open: [],
+      message: [],
+    };
+
+    this.open();
+  }
+
+  open() {
+    if (this.closing) {
+      return;
+    }
+    this.opening = true;
+    this.ws = new WebSocket(this.href);
+    this.ws.addEventListener('open', this.onOpen, {once: true});
+    this.ws.addEventListener('message', this.onMessage);
+    this.ws.addEventListener('close', this.reopen);
+    this.ws.addEventListener('error', this.reopn);
+  }
+
+  /**
+   * @param {string} name - Name of event to listen for
+   * @param {Function} listener - Called when underlying websocket emits event
+   */
+  addEventListener(name, listener) {
+    if (name !== 'open' && name != 'message') {
+      throw new Error(`Event ${name} not supported`);
+    }
+
+    if (name === 'open' && !this.opening) {
+      // readyState would cause race condition
+      listener();
+    }
+
+    this.listeners[name].push(listener);
+  }
+
+  onOpen() {
+    this.backoff = 1000;
+    this.opening = false;
+    for (const listener of this.listeners.open) {
+      listener();
+    }
+  }
+
+  onMessage(event) {
+    for (const listener of this.listeners.message) {
+      listener(event);
+    }
+  }
+
+  reopen() {
+    this.ws.removeEventListener('message', this.onMessage);
```

Comment:
Is there ever a case that this would be hit before `onOpen()` is called? If so, you should also remove the `open` listener.

Reply:
Yeah, I'll improve this logic since I haven't seen anything saying the readyState will always pass through open during some point of the WS's lifecycle